### PR TITLE
Glitter Ores Map Generation

### PR DESCRIPTION
**Pull request by TWLTriston**
Randomizes the ore patches outside of spawn area to density settings

### CODE REVIEW

**Comment by TWLTriston** [comment]
+1 sprinkle, addresses issue #19

File: locale/gen_ores/glitter_ores.lua [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+         mix_current = mix_current + 1
+         ore_mix[mix_current] = a
+      end
+   end
+      
+end
+
+--generate ores for entire chunk
+function run_ores_module(event)
+   run_ores_module_setup()
```
Comment (Valansch) [inline locale/gen_ores/glitter_ores.lua]: Why do you call the setup on every chunk load? This means doing the loop line 23 to 28 5000 times :O on every chunk generation.

Comment (TWLTriston) [reply locale/gen_ores/glitter_ores.lua]: I found that without the setup the savedscenarios didn't trigger the array initialization
i had a check in place for global.glitter_setup = false

But that gets set to true, and upon reloading the saved scenario, the ores weren't generating

Maybe another way to check for initialization, as globals persist but locals do not?

File: locale/gen_ores/glitter_ores.lua [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+-- Sets the buffer distance before ores are scrambled
+local starting_buffer = 100
+
+function run_ores_module_setup()
+	if glitter_debug then game.print("Glitter Ore: debug is enabled; module setup;") end
+
+   -- Prime the array
+   for a,b in pairs(ore_ratios) do
+      for i=1,(b*1000) do
+         mix_current = mix_current + 1
```
Comment (Valansch) [inline locale/gen_ores/glitter_ores.lua]: Mix_current will cause desyncs here. 
The following scenario:
- Server starts
- Server generates first chunk on tick 0 and set mix_current to some high number n.
- User joins game on tick 50. He loads the script and sets mix_current to 1 on line 6.
- 2nd chunk is generated on tick 64, both run this line.
- The clients result is 1+1 = 2, while the servers result is n + 1. The two have conflicting states and the client desyncs from the game.

File: locale/gen_ores/glitter_ores.lua (hunk 1 above)
Comment (Valansch) [reply locale/gen_ores/glitter_ores.lua]: Just run it once and store the data in a global table. That will solve both the desync and init problem.

File: locale/gen_ores/glitter_ores.lua (hunk 2 above)
Comment (TWLTriston) [reply locale/gen_ores/glitter_ores.lua]: Ok --- suggestion on how to build the array without desync?

Goal is to have an upper and lower bound to compare the random number against -- depending on outcome, was looking for an O[1] solution, but given we are only 4 elements really max, doing it O[n] may not be so bad.

Thoughts?

**Comment by TWLTriston** [comment]
Additional fixes published, please review again

File: map_layout.lua [hunk 3]
```diff
@@ -34,13 +34,13 @@ in this file and your run_*type*_module(event) function will be called.
 
 --ores--
 --require "locale.gen_ores.neko_crazy_ores"
+--require "locale.gen_ores.fluffy_rainbows"
+require "locale.gen_ores.glitter_ores"
```
Comment (Valansch) [inline map_layout.lua]: uncomment please :P

File: locale/gen_ores/glitter_ores.lua [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+      ["iron-ore"] = 1.0,
+      ["coal"] = 0.5,
+      ["copper-ore"] = 1,
+      ["stone"] = 0.25
+   }
+   -- 1-100% chance of sprinkling any individual ore
+   sprinkle_factor = 100
+
+   -- Sets the buffer distance before ores are scrambled
+   starting_buffer = 125
```
Comment (Valansch) [inline locale/gen_ores/glitter_ores.lua]: *noob buffer xD

File: locale/gen_ores/glitter_ores.lua [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+               local position_old = entity.position
+               local amount_old = entity.amount
+
+               local surface = entity.surface
+               entity.destroy()
+               local new_entity = surface.create_entity{name = new_name, position = position_old, force="neutral", amount=amount_old}
+            end
+         else
+            -- game.print("Skipping")
+            -- game.print(entity.name)
```
Comment (Valansch) [inline locale/gen_ores/glitter_ores.lua]: Can you remove any commented out code?

File: locale/gen_ores/glitter_ores.lua [hunk 6]
```diff
@@ -0,0 +1,76 @@
+-- Glittery ores, provide a mix value, and all patches outside uranium will be a full mix.
+glitter_debug = false
+
+function run_ores_module_setup()
+	if glitter_debug then game.print("Glitter Ore: debug is enabled; module setup;") end
```
Comment (Valansch) [inline locale/gen_ores/glitter_ores.lua]: One too many tabs spaces :O

File: locale/gen_ores/glitter_ores.lua [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+   
+   ore_mix_max = 0
+   -- Prime the array
+   for a,b in pairs(ore_ratios) do
+      for i=1,(b*1000) do
+         ore_mix_max = ore_mix_max + 1
+         ore_mix[ore_mix_max] = a
+      end
+   end
+
```
Comment (Valansch) [inline locale/gen_ores/glitter_ores.lua]: Extra line

File: locale/gen_ores/glitter_ores.lua (hunk 4 above)
Comment (TWLTriston) [reply locale/gen_ores/glitter_ores.lua]: Configurable!

**Comment by Valansch** [comment]
Looking good, ill merge. Thank you for your work :+1:

**Pull request merged**